tests/executors/test_local_executor.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from airflow.executors.local_executor import LocalExecutor
+from airflow.utils.state import State
+from airflow.utils.timeout import timeout
+
+
+class LocalExecutorTest(unittest.TestCase):
+
+    TEST_SUCCESS_COMMANDS = 5
+
+    def execution_parallelism(self, parallelism=0):
+        executor = LocalExecutor(parallelism=parallelism)
+        executor.start()
+
+        success_key = 'success {}'
+        success_command = 'echo {}'
+        fail_command = 'exit 1'
+
+        for i in range(self.TEST_SUCCESS_COMMANDS):
+            key, command = success_key.format(i), success_command.format(i)
+            executor.execute_async(key=key, command=command)
+            executor.running[key] = True
+
+        # errors are propagated for some reason
+        try:
+            executor.execute_async(key='fail', command=fail_command)
+        except:
+            pass
+
+        executor.running['fail'] = True
+
+        if parallelism == 0:
+            with timeout(seconds=5):
```

Comment:
Agreed. AFAIK this introduces a timeout using an external construct; it would be nice to have a timeout parameter in `BaseExecutor#end()` and `BaseExecutor#sync()` to avoid this patterns.